src/EditorFeatures/Core/Tagging/AsynchronousTaggerProvider.cs

```diff
@@ -3,79 +3,189 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
-using System.Text;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Options;
 using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Microsoft.VisualStudio.Text;
 using Microsoft.VisualStudio.Text.Editor;
 using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Editor.Tagging
 {
     /// <summary>
-    /// Computes <see cref="ITag"/>s for a text buffer in an asynchronous manner.  Computation of
-    /// the tags is handled by a provided <see cref="IAsynchronousTaggerDataSource{TTag}"/>.  The
-    /// <see cref="AsynchronousTaggerProvider{TTag}"/> handles the jobs of scheduling when to 
-    /// compute tags, managing the collection of tags, and notifying the and keeping the user 
-    /// interface up to date with the latest tags produced.
+    /// Base type of all asynchronous tagger providers (<see cref="ITaggerProvider"/> and <see cref="IViewTaggerProvider"/>). 
     /// </summary>
-    internal class AsynchronousTaggerProvider<TTag> : ITaggerProvider
+    internal abstract partial class AsynchronousTaggerProvider<TTag> : IAsynchronousTaggerDataSource<TTag>, ITaggerProvider, IViewTaggerProvider
         where TTag : ITag
     {
-        private readonly AsynchronousTaggerProviderImpl _taggerImplementation;
+        private readonly object uniqueKey = new object();
+        private readonly IAsynchronousOperationListener asyncListener;
+        private readonly IForegroundNotificationService notificationService;
 
-        /// <summary>
-        /// Creates a new <see cref="AsynchronousTaggerProvider{TTag}"/> using the provided 
-        /// <see cref="IAsynchronousTaggerDataSource{TTag}"/> to determine when to compute tags 
-        /// and to produce tags when appropriate.
-        /// </summary>
-        public AsynchronousTaggerProvider(IAsynchronousTaggerDataSource<TTag> dataSource)
-            : this(null, null, dataSource)
+        public virtual TaggerTextChangeBehavior TextChangeBehavior => TaggerTextChangeBehavior.None;
+        public virtual TaggerCaretChangeBehavior CaretChangeBehavior => TaggerCaretChangeBehavior.None;
+        public virtual SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
+
+        public virtual IEqualityComparer<TTag> TagComparer => null;
+
+        public virtual IEnumerable<Option<bool>> Options => null;
```

Comment:
Can you document what specifying these do? And can we use empty enumerables versus magic nulls, since null is evil?